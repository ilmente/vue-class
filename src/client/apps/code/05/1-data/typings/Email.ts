export enum EmailStatus {
    UNREAD = 'unread',
    READ = 'read',
    SPAM = 'spam',
}

export interface EmailInfo {
    id: string;
    sender: string;
    subject: string;
    timestamp: string;
    status: EmailStatus;
}

export interface Email extends EmailInfo {
    content: string;
}
