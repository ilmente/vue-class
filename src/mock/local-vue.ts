import { jest } from '@jest/globals';
import { createLocalVue } from '@vue/test-utils';
import VueRouter from 'vue-router';
import Vuex, { StoreOptions } from 'vuex';
import { EmailListState } from '@05/2-connection/modules/EmailListState';
import { EmailState } from '@05/2-connection/modules/EmailState';
import { MessagingState } from '@05/2-connection/modules/MessagingState';
import { staticInfoArray, staticSingleEmail } from './email';
import merge from 'lodash/merge';

export const localVueWithRouter = createLocalVue();
localVueWithRouter.use(VueRouter);
localVueWithRouter.use(Vuex);

export const localVue = createLocalVue();
localVue.use(Vuex);

export const router = new VueRouter();

export const generateStoreContext = (customOptions: StoreOptions<unknown> = {}) => {
    const defaultOptions = {
        modules: {
            emailList: {
                namespaced: true,
                state: new EmailListState,
                getters: {
                    filteredList: () => staticInfoArray,
                },
                actions: {
                    loadAll: jest.fn(),
                    setSpamOnly: jest.fn(),
                },
            },
            email: {
                namespaced: true,
                state: {
                    ...new EmailState,
                    current: staticSingleEmail,
                },
                getters: {
                    exists: () => true,
                    content: () => 'Content',
                },
                actions: {
                    load: jest.fn(),
                    updateEmailStatus: jest.fn(),
                },
            },
            messaging: {
                state: new MessagingState,
                actions: {
                    setLoading: jest.fn(),
                    push: jest.fn(),
                },
            }
        }
    };

    const options = merge(defaultOptions, customOptions);

    return {
        store: new Vuex.Store(options),
        options,
    }
};
