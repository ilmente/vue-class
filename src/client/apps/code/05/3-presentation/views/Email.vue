<template>
    <div class="email p-6">
        <div v-if="isLoading" class="email__progress">
            <progress 
                class="progress is-small is-primary" 
                max="100">
            </progress>
        </div>
        <article v-else-if="email">
            <StatusDropdown 
                class="is-pulled-right"
                :isUpdating="isUpdating"
                :value="email.status"  
                @beforeChange="onStatusBeforeChange"
                @change="onStatusChange" />

            <h1 class="title">{{ email.subject }}</h1>
            <h2 class="subtitle">
                <a>{{ email.sender }}</a>, {{ timestamp }}
            </h2>
            <hr>
            <div class="is-size-4 content" v-html="content"></div>
        </article>
        <div v-else class="email__error">
            <span class="email__icon">ðŸ¤¬ 500</span>
        </div>
    </div>
</template>

<style lang="scss">
    @use 'variables';

    .email {
        position: relative;
        height: 100%;
        overflow: auto;

        &__progress,
        &__error {
            position: relative;
            display: flex;
            align-items: center;
            justify-content: center;
            height: 100%;
        }

        &__icon {
            font-size: 10rem;
        }
    }
</style>

<script lang="ts">
    import { Component, Prop, Watch } from 'vue-property-decorator';
    import { Email, EmailStatus } from '../../1-data/typings/Email';;
    import StoreComponent from '../components/StoreComponent';
    import { getRelativeDate } from '../helpers/date';
    import StatusDropdown from '../components/StatusDropdown.vue';

    const DEFAULT_TIMEOUT = 1500; 

    @Component({
        name: 'Email',
        components: {
            StatusDropdown,
        }
    })
    export default class extends StoreComponent {
        @Prop({ type: String, required: true }) emailId!: string;
        isUpdating: boolean = false;
        timeoutId: ReturnType<typeof setTimeout> | null = null;

        get isLoading(): boolean {
            return this.$store.state.messaging.isLoading && !this.isUpdating;
        }

        get email(): Email | null {
            if (!this.$store.getters['email/exists']) {
                return null;
            }
            
            return this.$store.state.email.current;
        }

        get content(): string {
            return this.$store.getters['email/content'];
        }

        get timestamp(): string {
            return getRelativeDate(this.$store.state.email.current?.timestamp);
        }

        async created(): Promise<void> {
            await this.$store.dispatch('email/load', this.emailId);
            await this.autoUpdateStatus();
        }

        onStatusBeforeChange(): void {
            this.clearAutoUpdateStatus();
        }

        async onStatusChange(status: EmailStatus): Promise<void> {
            this.clearAutoUpdateStatus();
            
            await this.updateStatus(status);
        }

        @Watch('$route')
        async onRouteChange(): Promise<void> {
            this.clearAutoUpdateStatus();

            await this.$store.dispatch('email/load', this.emailId);
            await this.autoUpdateStatus();
        }

        clearAutoUpdateStatus(): void {
            if (!this.timeoutId) {
                return;
            }

            clearTimeout(this.timeoutId);
        }

        async autoUpdateStatus(): Promise<void> {
            if (this.email?.status !== EmailStatus.UNREAD) {
                return;
            }

            this.timeoutId = setTimeout(async () => {
                await this.updateStatus(EmailStatus.READ)
            }, DEFAULT_TIMEOUT);
        }

        async updateStatus(status: EmailStatus): Promise<void> {
            this.isUpdating = true;

            await this.$store.dispatch('email/updateEmailStatus', {
                id: this.email?.id, 
                status,
            });

            this.isUpdating = false;
        }
    }
</script>
