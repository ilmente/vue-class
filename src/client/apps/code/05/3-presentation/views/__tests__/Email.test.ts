import { mount } from '@vue/test-utils'
import Email from '../Email.vue';
import { generateStoreContext, localVueWithRouter, router } from '@mock/local-vue';

const propsData = {
    emailId: '123',
}

describe('Email.vue', () => {
    let context: any;

    const getWrapper = () => mount(Email, {
        localVue: localVueWithRouter,
        router,
        store: context.store,
        propsData,
    });
    
    beforeEach(() => {
        context = generateStoreContext();
    });

    test('snapshot', () => {
        const wrapper = getWrapper();
        expect(wrapper.element).toMatchSnapshot();
    });

    test('call load on created', () => {
        getWrapper();
        expect(context.options.modules.email.actions.load).toHaveBeenCalledTimes(1);
    });

    test('call load on created and on route change (2 times)', async (done) => {
        const wrapper = getWrapper();
        wrapper.vm.$router.push('/email/2');

        // this will not work
        // expect(context.options.modules.email.actions.load).toHaveBeenCalledTimes(2);

        wrapper.vm.$nextTick(() => {
            expect(context.options.modules.email.actions.load).toHaveBeenCalledTimes(2);
            done();
        });
    });
});
