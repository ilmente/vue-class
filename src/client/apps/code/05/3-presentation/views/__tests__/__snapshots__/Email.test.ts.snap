// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Email.vue snapshot 1`] = `
<div
  class="email p-6"
>
  <article>
    <div
      class="dropdown is-right is-pulled-right"
    >
      <div
        class="dropdown-trigger"
      >
        <button
          class="button is-info is-light"
          data-qa-status-dropdown-button=""
        >
          <span
            data-qa-status-dropdown-value=""
          >
            Read
          </span>
        </button>
      </div>
       
      <div
        class="dropdown-menu"
        id="dropdown-menu"
        role="menu"
      >
        <div
          class="dropdown-content"
        >
          <a
            class="dropdown-item"
            data-qa-status-dropdown-option=""
          >
            
                Unread
            
          </a>
          <a
            class="dropdown-item is-active"
            data-qa-status-dropdown-option=""
          >
            
                Read
            
          </a>
          <a
            class="dropdown-item"
            data-qa-status-dropdown-option=""
          >
            
                Spam
            
          </a>
        </div>
      </div>
    </div>
     
    <h1
      class="title"
    >
      Gorgeous Rubber Chicken
    </h1>
     
    <h2
      class="subtitle"
    >
      <a>
        Nikki_Brekke@example.org
      </a>
      , last Wednesday at 8:30 PM
        
    </h2>
     
    <hr />
     
    <div
      class="is-size-4 content"
    >
      Content
    </div>
  </article>
</div>
`;
